### PR TITLE
feat(1482): add collapse feature

### PR DESCRIPTION
**Pull request by minz1027**
- collapse waiting builds and re-enqueue the current build if it is the latest one
- add a feature flag for the collapse feature

`Related to`: https://github.com/screwdriver-cd/screwdriver/issues/1482

Set the feature flag to `false` in our worker, will enable that and test it when this is merged.

### CODE REVIEW

File: lib/BlockedBy.js [hunk 1]
```diff
@@ -69,6 +117,11 @@ class BlockedBy extends NodeResque.Plugin {
         const enforceBlockedBySelf = String(this.options.blockedBySelf) === 'true'; // because kubernetes value is a string
         const shouldDelete = await this.queueObject.connection.redis.get(deleteKey);
         const runningBuildId = await this.queueObject.connection.redis.get(runningKey);
+        const enableCollapse = String(this.options.collapse) === 'true'; // because kubernetes value is a string
+        const buildconfig = await this.queueObject.connection.redis.hget(
```
Comment (tkyi) [inline lib/BlockedBy.js]: camel case might be better
```suggestion
        const buildConfig = await this.queueObject.connection.redis.hget(
```

File: lib/BlockedBy.js [hunk 2]
```diff
@@ -114,7 +167,14 @@ class BlockedBy extends NodeResque.Plugin {
 
             // If any blocking job is running, then re-enqueue
             if (blockingBuildIds.length > 0) {
-                await this.reEnqueue(waitingKey, buildId, blockingBuildIds);
+                if (enforceBlockedBySelf && enableCollapse && collapse) {
```
Comment (tkyi) [inline lib/BlockedBy.js]: if `enforceBlockedBySelf` and `enableCollapse`/`collapse` need to be true for collapse to happen?
so if `enforceBlockedBySelf` is false and `enableCollapse`/`collapse` are true, we will still never see collapse behavior?

File: package-lock.json [hunk 3]
```diff
@@ -22,30 +22,24 @@
       }
     },
     "@sinonjs/formatio": {
-      "version": "3.0.0",
-      "resolved": "https://registry.npmjs.org/@sinonjs/formatio/-/formatio-3.0.0.tgz",
-      "integrity": "sha512-vdjoYLDptCgvtJs57ULshak3iJe4NW3sJ3g36xVDGff5AE8P30S6A093EIEPjdi2noGhfuNOEkbxt3J3awFW1w==",
+      "version": "3.1.0",
```
Comment (d2lam) [inline package-lock.json]: Can we add `package-lock.json` to `.gitignore`

File: lib/BlockedBy.js (hunk 2 above)
Comment (d2lam) [reply lib/BlockedBy.js]: I guess if blockedBy is not enabled, there will be no `waitingKeys`...

File: lib/BlockedBy.js [hunk 4]
```diff
@@ -2,7 +2,55 @@
...
+const winston = require('winston');
+
+/**
+ * collapse waiting builds and re-enequeue the current build if it is the latest one
+ * @method collapseBuilds
+ * @param  {String}  waitingKey           ${waitingJobsPrefix}${jobId}
+ * @param  {Number}  buildId              Current build Id
+ * @param  {Array}   blockingBuildIds     List of build Ids that are blocking this current build
+ */
+async function collapseBuilds({ waitingKey, buildId, blockingBuildIds }) {
```
Comment (d2lam) [inline lib/BlockedBy.js]: I think this logic might be simpler if we call it after `blockedBySelf`. That way `blockedBySelf` will do the adding to `waitingKey`, and `collapseBuilds` only has the responsibility to remove extra keys except for the last one.

File: lib/BlockedBy.js (hunk 1 above)
Comment (minz1027) [reply lib/BlockedBy.js]: fixed!

File: lib/BlockedBy.js (hunk 2 above)
Comment (minz1027) [reply lib/BlockedBy.js]: yeah, if it's not enforceBlockedBySelf, there won't be waiting builds to collapsed

File: package-lock.json (hunk 3 above)
Comment (minz1027) [reply package-lock.json]: removed it and add it to .gitignore

File: lib/BlockedBy.js (hunk 4 above)
Comment (minz1027) [reply lib/BlockedBy.js]: discussed offline, putting it after blockedBySelf may result in re-enqueue of the older builds

File: lib/BlockedBy.js (hunk 2 above)
Comment (tkyi) [reply lib/BlockedBy.js]: okay 👍 

we should call this out in the guide documentation

**Pull request merged**